Hw4

watchbot23 opened the pull request.

watchbot23 committed:
HW4
Task1
Task2

watchbot23 committed:
Rewrited class Worker (added cycles)

watchbot23 committed:
some little fixes

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,143 @@
+        static void Main(string[] args)
+        {
+            const string laptop = "Laptop";
+            const string server = "Server";
```
Put these string inside classes. They should now who they are

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,143 @@
+    }
+    class LapTop
+    {
+        public string Name;
```
Actually you need to protect these fields from rewriting from outside. Anyway you pass info to constructor.

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,143 @@
+    {
+        public string Name;
+        public string Power;
+        public string MP3;
```
Bad solution. You hardcoded types. You should create one field/property for types.

LiDmYr replied:
Create additional class for type

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Program.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        static void Main(string[] args)
+        {
+            Worker anton = new Worker();
```
It's not flexible solution for handling workers. At least use an array to handle workers

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Program.cs:
```diff
@@ -0,0 +1,29 @@
+
+            for (int i = 0; agregat.Length < completeAgrigat; i++)
+            {
+                agregat = anton.JuniorWorker(agregat);
```
Hardcoded solution. You shouldn't call specific method. You should call "DoWork" and every worker should do job according to his skills.

LiDmYr replied:
Additionaly create class Agregat

LiDmYr reviewed (changes requested):
Please implement required changes according my comments

Commit:
Created new class Agregat
Little chenges in Task1

watchbot23 committed:
added class Agregat
added DoWork() for class Worker
added object array for Workers

watchbot23 committed:
Task 3 (just a little part)

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+    }
+    class LapTop
+    {
+        public string laptop = "Laptop";
```
anybody from outside can change the value

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+    class LapTop
+    {
+        public string laptop = "Laptop";
+        protected string Name;
```
wrong solution to protect your variables. Read again about Access modifiers.
 delete protected modifiers everywhere in your solution.

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+        {
+            LapTop Acer = new LapTop("Acer", 65, 8, 2.1);
+            LapTop MSI = new LapTop("MSI", 85, 16, 2.9);
+            Console.WriteLine(Acer.laptop + " 1:\n" + Acer.GetDescription());
```
**Acer.GetDescription()** should give all info. Don't call separately **Acer.laptop**

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+    {
+        static void Main(string[] args)
+        {
+            LapTop Acer = new LapTop("Acer", 65, 8, 2.1);
```
local variables start with lowerCase

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+            Name = name;
+            Power = $"{power} W";
+            Memory = $"{memory} Gb";
+            Waight = $"{waight} Kg";
```
w**E**ight

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+    }
+    class PlasmTV
+    {
+        public string plazmTV = "PlasmTV";
```
rename **plazmTV** to something else which describes the meaning of the variable's context. Change the same everywhere

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+            return $"  Name: {Name} {Environment.NewLine}  Power: {Power} {Environment.NewLine}  ScreenDiagonal: {ScreenDiagonal} {Environment.NewLine}  Frequency: {Frequency}";
+        }
+    }
+    class PleyerDevice
```
Pl**A**yer

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+        public string pleyer = "Pleyer Device";
+        protected string Name;
+        protected string Power;
+        protected string KodekSupported;
```
Add new class for **Format** ( inside should be something which describes - name of format, bitreit of format). 
Class **Player** should include collection of formats.

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Agregat.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    class Agregat
+    {
+        protected string AgregatValue = "Agregat";
```
I don't understand what is this...

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Agregat.cs:
```diff
@@ -0,0 +1,17 @@
+namespace GDM.HW4.OOP.Classes.Task2
+{
+    class Agregat
+    {
```
Please create class Aggregate which completely describe the business model (you need to thing about this). Now it's actually nothing.

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Program.cs:
```diff
@@ -0,0 +1,38 @@
+        static void Main(string[] args)
+        {
+            int workersQuantity = 3;
+            string[] workersNames = new string[3] { "Anton", "Bogdan", "Clark" };//примитивная симуляция базы данных
```
Don't write useless comments. Write understandable code

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Program.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            int workersQuantity = 3;
+            string[] workersNames = new string[3] { "Anton", "Bogdan", "Clark" };//примитивная симуляция базы данных
+            int[] workersXP = new int[3] { 1, 3, 7 };//примитивная симуляция базы данных
```
create method which is preparing workers

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Program.cs:
```diff
@@ -0,0 +1,38 @@
+            }
+            Agregat agregatValue = new Agregat();
+            string agregat = agregatValue.GetAgregatValue();
+            int completeAgrigat = 50;
```
wrong place for this data. Bad OOP provides problems with data location

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Program.cs:
```diff
@@ -0,0 +1,38 @@
+                workers[i].Expirience = workersXP[i];
+            }
+            Agregat agregatValue = new Agregat();
+            string agregat = agregatValue.GetAgregatValue();
```
I don't uderstand the context and purpose of this variable

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Program.cs:
```diff
@@ -0,0 +1,38 @@
+
+            for (int i = 0; agregat.Length < completeAgrigat; i++)
+            {
+                agregat = workers[0].DoWork(agregat, completeAgrigat, workers[0].Expirience);
```
Imagint you have 1000 workers....Will you write 1000 time this line? Of course, no - do refactoring

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Program.cs:
```diff
@@ -0,0 +1,38 @@
+            {
+                agregat = workers[0].DoWork(agregat, completeAgrigat, workers[0].Expirience);
+                agregat = workers[1].DoWork(agregat, completeAgrigat, workers[1].Expirience);
+                agregat = workers[2].DoWork(agregat, completeAgrigat, workers[2].Expirience);
```
Worker should know about his experience. Don't pass him this info.

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Worker.cs:
```diff
@@ -0,0 +1,45 @@
+    class Worker
+    {
+        string[] WorkersXP = new string[3] { "Junior", "Middle", "Senior" };
+        public string Name { get; set; }
```
make "private set". Read about this ot watch the video from class

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Worker.cs:
```diff
@@ -0,0 +1,45 @@
+{
+    class Worker
+    {
+        string[] WorkersXP = new string[3] { "Junior", "Middle", "Senior" };
```
Delete this line. Why does worker have property with all experiences ?

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Worker.cs:
```diff
@@ -0,0 +1,45 @@
+    {
+        string[] WorkersXP = new string[3] { "Junior", "Middle", "Senior" };
+        public string Name { get; set; }
+        public int Expirience { get; set; }
```
Exp**E**rience

LiDmYr commented on GDM.HW4.OOP.Classes.Task2/Worker.cs:
```diff
@@ -0,0 +1,45 @@
+        public int Expirience { get; set; }
+
+        public string DoWork(string agregat, int completeAgrigat, int workersXPYears)
+        {
```
I don't understand what is going on inside this method

LiDmYr reviewed (changes requested):
Need works

Commit:
Little sintax fixes in Task1
Complete Task3

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+namespace GDM.HW4.OOP.Classes.Task3
+{
+    class Program
+    {
```
Class Program should not have these fields

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+    {
+        private static Shop _shop;
+        private static MobileStore[] _mobStore;
+        private static Phone[][] _phone;
```
Phones are a part of mobile store

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+    class Program
+    {
+        private static Shop _shop;
+        private static MobileStore[] _mobStore;
```
stores are a part of shop

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+        {
+            Console.WriteLine("- - - - - NEW COMMAND - - - - -");
+            Console.WriteLine("Please write index of command from list below. Commands: " +
+                "\n [0] - quit " +
```
Imagine that you get commands from some method. Don't hardcode 28-32 lines. Do refactoring

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+                "\n [4] - clear console");
+
+            string response = "";
+            response = Console.ReadLine();
```
Combine 34+35 lines

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+
+            string response = "";
+            response = Console.ReadLine();
+            if (response == "")
```
Use  **string.IsNullOrWhiteSpace** for this purpose

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+            if (response == "")
+            {
+                Console.WriteLine("Invalid command");
+                ConsoleOptionMenu();
```
NO RECURSION IN THIS CASE AND IN THIS TASK AT ALL!!!! ACHTUNG!!!!

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+
+            Console.ReadLine();
+        }
+        public static void ConsoleOptionMenu ()
```
You named method like "Print in console menu option" but you method is doing everything for your shop. Please create new name and rewrite completely method and logic. And please think always what is a name method and what actually method is doing inside!!!

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+                }
+            }
+        }
+        public static bool IsDigitsOnly(string consoleString)
```
Use Int.TryParse instead of this method. You wrote your own bicycle

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+            for (int i = 0; i < _mobStore.Length; i++)
+            {
+                string storeCellStatus = "";
+                if (IsObjectInArrayNull(_mobStore, i) == true)
```
ACHTUNG!!!! we don't compare booleans. 
Write: **if(bool){...}**  instead of **if(bool == bool){...}**

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+                    for (int j = 0; j < _mobStore[i].Capasity; j++)
+                    {
+                        string phoneCellStatus = "";
+                        if (IsObjectInArraysArrayNull(_phone, i, j) == true)
```
133-141. You have to call ConsoleWriteLine just once...rewrite

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Program.cs:
```diff
@@ -0,0 +1,371 @@
+        }
+        public static bool IsObjectInArrayNull (MobileStore[] array, int index)
+        {
+            bool tempBool = false;
```
terrible variable's name

LiDmYr reviewed (changes requested):
Unfortunately, it seems you don't understand recursion purpose - please don't use it at all for some time. 
I used just do {... } while () for doing that what you made with recursion and now you don't see disadvantages of chosen approach.
Do refactoring of your code according my comments and we will back later to your code.
Anyway, great job! Your console UI has required behaviour!!

watchbot23 committed:
- Removed comparing of bool with bool
- Rewrited logic in SetNewPhoneModelNameAndPrice() (bag fixed)

watchbot23 committed:
Refactored Class Program

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+    }
+    class PlayerDevice
+    {
+        public string player = "Pleyer Device";
```
`   public const string DeviceType = "Player Device";`

LiDmYr commented on GDM.HW4.OOP.Classes.Task1/Program.cs:
```diff
@@ -0,0 +1,140 @@
+    }
+    class LEDTV
+    {
+        public string ledTV = "LED TV";
```
`   public const string DeviceType = "LED TV";`

watchbot23 committed:
Refactored:
- Added Classes fields in Classes
- Removed recursion from all methods
- Created new method PrintOptionsMenu()

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Shop.cs:
```diff
@@ -0,0 +1,75 @@
+{
+    public class Shop
+    {
+        private string _name;
```
имя поля и проперти должны совпадать

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Shop.cs:
```diff
@@ -0,0 +1,75 @@
+    public class Shop
+    {
+        private string _name;
+        private int _numberOfStores;
```
поля должны быть обьявлены в упор к проперти, которые к ним привязаны

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Shop.cs:
```diff
@@ -0,0 +1,75 @@
+            {
+                if (value.Length < 1)
+                {
+                    Console.WriteLine("Please write valid (not empty) name for shop");
```
Модель привязана к консоли, но она может быть использована в Веб, Десктоп проектах - консоль убрать

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Shop.cs:
```diff
@@ -0,0 +1,75 @@
+        /// </summary>
+        public void SetNumberOfStores()
+        {
+            string storesNumber;
```
Класс Шоп не должен выполнять никакой логики связанной с взаимодействием с Юзером через консоль!!!

LiDmYr commented on GDM.HW4.OOP.Classes.Task3/Shop.cs:
```diff
@@ -0,0 +1,75 @@
+                }
+            }
+        }
+        public MobileStore MobStores { get; set; }
```
эм......
```
        public MobileStore MobStores { get; set; }
        public MobileStore[] MobStoresArray { get; set; }
```

это что?

LiDmYr reviewed (changes requested):
много функционального программирования и мало ООП - надо рефакторить код

Commit:
refactoring
